### PR TITLE
Add source locations to annotations.

### PR DESCRIPTION
**Pull request by smolkaj**
Proposal to address issue #285.

### CODE REVIEW

File: proto/p4/config/v1/p4types.proto [hunk 1]
```diff
@@ -181,6 +181,17 @@ message StructuredAnnotation {
     ExpressionList expression_list = 2;
```
Comment (antoninbas) [inline proto/p4/config/v1/p4types.proto:181]: there are quite a few messages in this file with an `annotations` field. Any reason why we are not adding `annotation_locations` for them (@chrispsommers I just noticed that this also applies for `structured_annotations`)?

Comment (smolkaj) [reply proto/p4/config/v1/p4types.proto:193]: Thanks for pointing that out, I missed those.
I also went ahead and added structured annotations where they were missing.

**Comment by smolkaj** [comment]
Yes, I still have to update the spec. Hope to get to it tomorrow.

File: proto/p4/config/v1/p4types.proto [hunk 2]
```diff
@@ -189,10 +212,14 @@ message P4EnumTypeSpec {
   message Member {
     string name = 1;
     repeated string annotations = 2;
+  // The location of `annotations[i]` is given by `annotation_locations[i]`.
+  repeated SourceLocation annotation_locations = 4;
```
Comment (chrispsommers) [inline proto/p4/config/v1/p4types.proto:216]: nit: indent this line and next?

File: proto/p4/config/v1/p4types.proto [hunk 3]
```diff
@@ -204,11 +231,15 @@ message P4SerializableEnumTypeSpec {
     string name = 1;
     bytes value = 2;
     repeated string annotations = 3;
+  // The location of `annotations[i]` is given by `annotation_locations[i]`.
```
Comment (chrispsommers) [inline proto/p4/config/v1/p4types.proto:234]: nit: indent this line and next?

File: proto/p4/config/v1/p4types.proto (hunk 2 above)
Comment (smolkaj) [reply proto/p4/config/v1/p4types.proto:216]: done

File: proto/p4/config/v1/p4types.proto (hunk 3 above)
Comment (smolkaj) [reply proto/p4/config/v1/p4types.proto:234]: done.

**Comment by smolkaj** [comment]
Finally got around to this.
I just ran the `update_go.sh` script, so if the CI passes we should be good to go.

**Comment by smolkaj** [comment]
Sweet, thanks. Indeed I did not check the rendering. Fingers crossed.

**Comment by smolkaj** [comment]
@antoninbas are we ready to merge then?

**Comment by antoninbas** [comment]
@smolkaj we are ready to merge as far as I'm concerned. Did anyone from your side (Konstantin / Stefan) want to review this?

File: docs/v1/P4Runtime-Spec.mdk [hunk 4]
```diff
@@ -1187,6 +1194,41 @@ structured_annotations {
...
 
+### `SourceLocation` Message
+
+A source location describes a location within a *.p4*-source file. The
+`SourceLocation` message is defined in p4types.proto as follows:
+
+~ Begin Proto
+// Location of code relative to a given source file.
+message SourceLocation {
+  // Path to the source file (total or relative to the working directory).
```
Comment (stefanheule) [inline docs/v1/P4Runtime-Spec.mdk:1205]: I think this is normally called absolute, not total.

Comment (stefanheule) [reply docs/v1/P4Runtime-Spec.mdk:1205]: Here and below.

Comment (smolkaj) [reply docs/v1/P4Runtime-Spec.mdk:1205]: Good catch. Done.

**Pull request merged**